source/lac/petsc_communication_pattern.cc

```diff
@@ -0,0 +1,516 @@
+                                const_cast<PetscInt *>(
+                                  inloc.size() > 0 ? inloc.data() : nullptr),
+                                PETSC_COPY_VALUES,
+                                remotes,
+                                PETSC_OWN_POINTER));
+    AssertPETSc(PetscSFSetUp(sf2));
+    // AssertPETSc(PetscSFView(sf2, nullptr));
+    // We need to invert root and leaf space to create the first SF
+    AssertPETSc(PetscSFCreateInverseSF(sf2, &sf1));
+    AssertPETSc(PetscSFDestroy(&sf2));
+    // AssertPETSc(PetscSFView(sf1, nullptr));
+
+    // Now create the SF from the contiguous space to the local output space
+    n = static_cast<PetscInt>(outidx.size());
+    AssertPETSc(PetscSFCreate(communicator, &sf2));
+    AssertPETSc(PetscSFSetGraphLayout(
+      sf2,
+      layout,
+      n,
+      const_cast<PetscInt *>(outloc.size() > 0 ? outloc.data() : nullptr),
+      PETSC_COPY_VALUES,
+      const_cast<PetscInt *>(n > 0 ? outidx.data() : nullptr)));
+    AssertPETSc(PetscSFSetUp(sf2));
+    // AssertPETSc(PetscSFView(sf2, nullptr));
+
+    // The final SF is the composition of the two
+    AssertPETSc(PetscSFCompose(sf1, sf2, &sf));
+    // AssertPETSc(PetscSFView(sf, nullptr));
+
+    // Cleanup
+    AssertPETSc(PetscLayoutDestroy(&layout));
+    AssertPETSc(PetscSFDestroy(&sf1));
+    AssertPETSc(PetscSFDestroy(&sf2));
+  }
+
+  void
+  CommunicationPattern::clear()
+  {
+    AssertPETSc(PetscSFDestroy(&sf));
+  }
+
+  const MPI_Comm &
+  CommunicationPattern::get_mpi_communicator() const
+  {
+    this->returncomm = PetscObjectComm(reinterpret_cast<PetscObject>(sf));
+    return this->returncomm;
+  }
+
+  template <typename Number>
+  void
+  CommunicationPattern::export_to_ghosted_array_start(
+    const ArrayView<const Number> &src,
+    const ArrayView<Number> &      dst) const
+  {
+    auto datatype = Utilities::MPI::mpi_type_id_for_type<Number>;
+
+#  if DEAL_II_PETSC_VERSION_LT(3, 15, 0)
+    AssertPETSc(PetscSFBcastBegin(sf, datatype, src.data(), dst.data()));
+#  else
+    AssertPETSc(
+      PetscSFBcastBegin(sf, datatype, src.data(), dst.data(), MPI_REPLACE));
+#  endif
+  }
+
+  template <typename Number>
+  void
+  CommunicationPattern::export_to_ghosted_array_finish(
+    const ArrayView<const Number> &src,
+    const ArrayView<Number> &      dst) const
+  {
+    auto datatype = Utilities::MPI::mpi_type_id_for_type<Number>;
+
+#  if DEAL_II_PETSC_VERSION_LT(3, 15, 0)
+    AssertPETSc(PetscSFBcastEnd(sf, datatype, src.data(), dst.data()));
+#  else
+    AssertPETSc(
+      PetscSFBcastEnd(sf, datatype, src.data(), dst.data(), MPI_REPLACE));
+#  endif
+  }
+
+  template <typename Number>
+  void
+  CommunicationPattern::export_to_ghosted_array(
+    const ArrayView<const Number> &src,
+    const ArrayView<Number> &      dst) const
+  {
+    export_to_ghosted_array_start(src, dst);
+    export_to_ghosted_array_finish(src, dst);
+  }
+
+  template <typename Number>
```

Comment:
Total aside: We generally have three empty lines between functions. No need to change anything here, eventually someone will clean it up, but if you want to keep it in mind for future patches :-)